### PR TITLE
Dashboard creation

### PR DESCRIPTION
**Pull request by pfafflabatuiuc**
First version of the dashboard/logger.

### CODE REVIEW

File: instrumentserver/dashboard/config.py [hunk 1]
```diff
@@ -0,0 +1,56 @@
+config = {
```
Comment (wpfff) [inline instrumentserver/dashboard/config.py:1]: this file looks like an example, right? 
maybe it'd be better make this more clear?
My favorite option would be to have an 'example' folder in the repository root directory.

File: instrumentserver/dashboard/dashboard.py [hunk 2]
```diff
@@ -0,0 +1,290 @@
...
+from bokeh.plotting import figure
+from bokeh.palettes import Category10
+
+from .config import config
+
+
+class PlotParameter:
+    """
+    PlotParameter is used to hold an individual parameter of the plot.
+    It saves the data of it as well as draws the lines for the figures.
```
Comment (wpfff) [inline instrumentserver/dashboard/dashboard.py:18]: is it still saving at this moment?

Comment (pfafflabatuiuc) [reply instrumentserver/dashboard/dashboard.py:18]: It was a typo, I corrected it.

**Comment by wpfff** [comment]
@marcosfrenkel -- it looks like we have a bunch of conflicts. it seems that adding a script to be able to launch the parameter manager as command is conflicting with the dashboard stuff. could you have a look and fix that?
The best way would be to merge the changes from the current master branch into your dashboard branch, and then push that branch again.

i'm also adding a few other comments here and there.

File: instrumentserver/testing/dummy_instruments/generic.py [hunk 3]
```diff
@@ -38,4 +39,49 @@ def __init__(self, name: str, *args, **kwargs):
             self.add_submodule(chan_name, channel)
 
     def test_func(self, a, b, *args, c: List[int] = [10, 11], **kwargs):
-        return a, b, args[0], c, kwargs['d'], self.param0()
\ No newline at end of file
+        return a, b, args[0], c, kwargs['d'], self.param0()
+
+class DummyInstrumentRandomNumber(Instrument):
```
Comment (wpfff) [inline instrumentserver/testing/dummy_instruments/generic.py:44]: I don't understand this instrument.
how are the random numbers set? do you have to call the generate_data function somewhere?
I would've expected simply that calling the parameter would generate random numbers?

File: setup.py [hunk 4]
```diff
@@ -9,5 +9,8 @@
       license='MIT',
       packages=['instrumentserver'],
       zip_safe=False,
-      entry_points={"console_scripts": ["instrumentserver = instrumentserver.apps:serverScript"]}
+      entry_points={"console_scripts": ["instrumentserver = instrumentserver.apps:serverScript",
```
Comment (wpfff) [inline setup.py:12]: these script names are very generic and it's not immediately clear that they belong to the instrumentserver package.
I would prefer naming them 'instrumentserver-dashboard', or something like that.

File: instrumentserver/dashboard/logger.py [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+
+    def update(self):
+        """
+        Updates the parameter and save a new data point in memory
+        """
+
+        # check that the source type is parameter
+        if self.source_type == 'parameter':
+
+            # just for development
```
Comment (wpfff) [inline instrumentserver/dashboard/logger.py:68]: this should not be in here, right?

File: instrumentserver/dashboard/config.py [hunk 6]
```diff
@@ -0,0 +1 @@
+config = {}
```
Comment (wpfff) [inline instrumentserver/dashboard/config.py:1]: do we need this file?

File: instrumentserver/dashboard/dashboard.py [hunk 7]
```diff
@@ -0,0 +1,281 @@
...
+from typing import Optional, List
+
+from bokeh.layouts import column, row
+from bokeh.models import ColumnDataSource, CheckboxGroup, DatetimeTickFormatter, HoverTool, DataRange1d
+from bokeh.models.widgets import Button, Tabs, Panel, Paragraph
+from bokeh.plotting import figure
+from bokeh.palettes import Category10
+
+
+from .__init__ import read_config
```
Comment (wpfff) [inline instrumentserver/dashboard/dashboard.py:13]: you don't need the __init__ part. just from . import read_config is fine.

File: instrumentserver/dashboard/__init__.py [hunk 8]
```diff
@@ -0,0 +1,257 @@
...
+
+    parameter_path: Full name of the parameter, including all of its submodules, as it appears on the instrument driver.
+
+    address: The address of the instrumentserver where the instrument of this parameters lives.
+
+How to set up config file:
+^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The file should be located in the dashboard folder inside of the instrumentserver package and should be called:
+'config.py'. The file should have a python dictionary named config, this is the dictionary that is being read.
```
Comment (wpfff) [inline instrumentserver/dashboard/__init__.py:34]: this is not good. 
the user should never have to put stuff into the package.

a solution could be to look for a config file in the current working directory (with a default name -- say 'instrumentserver-dashboard-cfg.py') and the user should be able to specify a filepath for one, alternatively.

Comment (wpfff) [reply instrumentserver/dashboard/__init__.py:34]: you can import from a specific file using importlib.util -- see for instance: https://codeigo.com/python/import-function-from-file#import-file-from-another-location

File: instrumentserver/dashboard/logger.py [hunk 9]
```diff
@@ -0,0 +1,175 @@
...
+
+
+class ParameterLogger(QtCore.QObject):
+    """
+    Main class of the logger. All of the parameters are saved inside this class
+    """
+    def __init__(self):
+        super().__init__()
+
+        # used for testing, the instruments should be already created for the dashboard to work
```
Comment (wpfff) [inline instrumentserver/dashboard/logger.py:88]: i'd prefer if the test wasn't hardcoded into this file. a better solution would be to have a separate testing script.

File: instrumentserver/dashboard/__init__.py [hunk 10]
```diff
@@ -0,0 +1,257 @@
...
+                [str] Directory of where the dashboard will load data. This should end with <name_of_file>.csv
+            load_and_save
+                [str] Directory of where the dashboard will load and the logger save data.
+                If this option is present, the dashboard will override save_directory and load_directory options.
+                This should end with <name_of_file>.csv
+
+Example:
+^^^^^^^^
+
+The following is an example dictionary::
```
Comment (wpfff) [inline instrumentserver/dashboard/__init__.py:92]: i think we should adapt the config structure a bit:
i think it makes more sense to have entries 'plots' and 'options' (i.e., add one more level of nesting for the plots). i think it's too confusing to have one particular name ('options') not be a plot.

File: instrumentserver/testing/dummy_instruments/generic.py (hunk 3 above)
Comment (pfafflabatuiuc) [reply instrumentserver/testing/dummy_instruments/generic.py:62]: I could not make this work by overriding the get function of this instrument. When I call it something else (not get) and hard-code the new name into the logger, the instrument works as intended. However if I leave the logger to use the generic qcodes instrument get function it goes straight into the Instrument get and not the overridden method. I could not figure out how to make it properly work. 

At the moment the logger and dashboard are technically functioning with this instrument, but it is not creating random numbers.

**Comment by pfafflabatuiuc** [comment]
I have addressed most reviews (check the only unresolved above). I could not get the random number dummy instrument to properly work but decided to stop wasting time on it and work on the rest of the pull request.  

I have the dashboard and logger running and I plan on leaving them to run overnight as the fridge cools down. Right now everything else seems to be working as expected.

**Comment by pfafflabatuiuc** [comment]
Just an update about the logger/dasher overnight test. The dashboard and logger run overnight logging 8 temperature channels and 2 pressure channels of the fridge. For all of them it gathered an update once every minute and in about 20 hours the file generated was a little bit less than a 1MB.

**Pull request merged**